cassiopeia/solver/PercolationSolver.py

```diff
@@ -0,0 +1,350 @@
+            in forming weights to scale the contribution of each mutation in
+            the similarity graph
+        similarity_function: A function that calculates a similarity score
+            between two given samples and their observed mutations. The default
+            is "hamming_distance_without_missing"
+        threshold: A minimum similarity threshold to include an edge in the
+            similarity graph
+
+    Attributes:
+        joining_solver: The Solver that is used to cluster groups of samples
+            after percolation steps that produce more than two groups
+        prior_transformation: Function to use when transforming priors into
+            weights.
+        similarity_function: A function that calculates a similarity score
+            between two given samples and their observed mutations
+        threshold: A minimum similarity threshold
+        
+    """
+
+    def __init__(
+        self,
+        joining_solver: CassiopeiaSolver.CassiopeiaSolver,
+        prior_transformation: str = "negative_log",
+        similarity_function: Optional[
+            Callable[
+                [
+                    List[int],
+                    List[int],
+                    int,
+                    Optional[Dict[int, Dict[int, float]]],
+                ],
+                float,
+            ]
+        ] = dissimilarity_functions.hamming_similarity_without_missing,
+        threshold: Optional[int] = 0,
+    ):
+
+        super().__init__(prior_transformation)
+
+        self.joining_solver = joining_solver
+        self.threshold = threshold
+        self.similarity_function = similarity_function
+
+
+    def solve(self,
+            cassiopeia_tree: CassiopeiaTree):
+        """Implements a solving procedure for the Percolation Algorithm.
+
+        The procedure recursively splits a set of samples to build a tree. At
+        each partition of the samples produced by the percolation procedure, 
+        an ancestral node is created and each side of the partition is placed 
+        as a daughter clade of that node. This continues until each side of 
+        the partition is comprised only of single samples. If an algorithm 
+        cannot produce a split on a set of samples, then those samples are 
+        placed as sister nodes and the procedure terminates, generating a 
+        polytomy in the tree. This function will populate a tree inside the 
+        input CassiopeiaTree.
+
+        Args:
+            cassiopeia_tree: CassiopeiaTree storing a character matrix and
+                priors.
+        """
+
+        # A helper function that builds the subtree given a set of samples
+        def _solve(
+            samples: List[Union[str, int]],
+            tree: nx.DiGraph,
+            character_matrix: pd.DataFrame,
+            unique_character_matrix: pd.DataFrame,
+            priors: Dict[int, Dict[int, float]],
+            weights: Dict[int, Dict[int, float]],
```

Comment:
You could just pass the priors and compute the weights over and over again (not sure if this would rack up compute time or not)

Reply:
Sure I'll do that!

Reply:
Actually I think I'll keep it in, as we pass it around too much and recalculating it at every step seems like it could be costly for large state distributions